### PR TITLE
Extend Chat Example w/ Gossip Based Backend

### PR DESCRIPTION
**Pull request by rohitkulshreshtha**
Added a gossip-based backend for the chat example.

The implementation is based on ["Epidemic algorithms for replicated database maintenance"](https://dl.acm.org/doi/epdf/10.1145/41840.41841). Specifically, it implements push-based "rumor-mongering" with a blind-coin removal process.

### CODE REVIEW

**Review by MingweiSamuel — COMMENTED** [review]
First thought is that it would be best to have separate regular vs gossip version of chat, what do you think?

**Review by jhellerstein — CHANGES_REQUESTED** [review]
Looks great! Can you augment the README to discuss the different roles and how they work?

**Comment by rohitkulshreshtha** [comment]
> First thought is that it would be best to have separate regular vs gossip version of chat, what do you think?

In one of the stand-ups, Joe suggested interop with the existing client as a goal. This seems like the easiest way to do that. What'd you have in mind?

File: hydroflow/examples/chat/main.rs [hunk 1]
```diff
@@ -23,9 +37,7 @@ struct Opts {
     #[clap(value_enum, long)]
     role: Role,
     #[clap(long, value_parser = ipv4_resolve)]
-    addr: Option<SocketAddr>,
-    #[clap(long, value_parser = ipv4_resolve)]
-    server_addr: Option<SocketAddr>,
```
Comment (MingweiSamuel) [inline hydroflow/examples/chat/main.rs:28]: Love this new simpler parameterization

File: hydroflow/examples/chat/main.rs [hunk 2]
```diff
@@ -2,18 +2,32 @@ use std::net::SocketAddr;
...
 
+use crate::randomized_gossiping_server::run_gossiping_server;
+
 mod client;
 mod protocol;
+mod randomized_gossiping_server;
 mod server;
 
-#[derive(Clone, ValueEnum, Debug)]
+#[derive(Clone, ValueEnum, Debug, Eq, PartialEq)]
```
Comment (MingweiSamuel) [inline hydroflow/examples/chat/main.rs:16]: Nit: Add `Copy` too

File: hydroflow/examples/chat/randomized_gossiping_server.rs [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+    let all_members = [
+        GossipingServer1,
+        GossipingServer2,
+        GossipingServer3,
+        GossipingServer4,
+        GossipingServer5,
+    ];
+
+    let other_members: Vec<Role> = all_members
+        .iter()
```
Comment (MingweiSamuel) [inline hydroflow/examples/chat/randomized_gossiping_server.rs:76]: nit: can use `into_iter()` instead and avoid `cloned()/copied()`

**Review by MingweiSamuel — APPROVED** [review]
Ok, sounds good to me then

**Comment by jhellerstein** [comment]
> > First thought is that it would be best to have separate regular vs gossip version of chat, what do you think?
> 
> In one of the stand-ups, Joe suggested interop with the existing client as a goal. This seems like the easiest way to do that. What'd you have in mind?

The current code supports letting readers see the distinction between single-server vs gossiping-server, and run either one from the command line. The README.md should talk that through in a bit more detail is all.

File: hydroflow/examples/chat/main.rs (hunk 1 above)
Comment (rohitkulshreshtha) [reply hydroflow/examples/chat/main.rs:28]: Created https://github.com/hydro-project/hydroflow/issues/1123 for follow-up.

**Comment by rohitkulshreshtha** [comment]
> The current code supports letting readers see the distinction between single-server vs gossiping-server, and run either one from the command line. The README.md should talk that through in a bit more detail is all.

This is done. The new readme can also be run straight from the IDE (at least IntelliJ at the moment, unsure what vscode does with `shell` code blocks.

**Review by jhellerstein — APPROVED** [review]
LGTM

**Comment by rohitkulshreshtha** [comment]
Only conventional commits didn't pass, so squashing and merging. Wondering if we should configure that check differently.

**Pull request merged**